### PR TITLE
fix: validate in hook-form rules

### PR DESCRIPTION
**Pull request by infothien**
aimed to fix this problem https://github.com/scoville/ats/issues/327

Now we can use validate in Rules of Hook-form 

https://user-images.githubusercontent.com/30365498/125015464-7fe97300-e0aa-11eb-8dab-ed8607dff168.mov

### CODE REVIEW

File: src/Rules.res [hunk 1]
```diff
@@ -1,7 +1,3 @@
-module Validation = {
-  type t = Sync(string => bool) | Async(string => Js.Promise.t<bool>)
```
Comment (vtrofin) [inline src/Rules.res:2]: While i do understand the fix with the union type i'm not sure i properly understand why type Validation.t failed. I used the Sync variant constructor with a string => bool function. What is the representation of this type at runtime?

**Review by vtrofin — APPROVED** [review]
Thank you @infothien , it's a nice solution to a problem i'm not sure I properly understand yet. I left an inline comment in case you or @gaku-sei might be able to explain it more in-depth.

Thank you very much for this fix!

File: src/Rules.res (hunk 1 above)
Comment (infothien) [reply src/Rules.res:2]: I made a [demo](https://rescript-lang.org/try?code=LYewJgrgNgpgBANQIZQJZiQF1SAdnAXjgG8AoOOTATwAd5NC4BlK3AYwAoBnTAJ1VwBzQgD44AIxAgoASjgAfOAEEurTj35DRcAFJcAdAAVeIYKi4x9mADyTpImaQC+pV3v1QQgjsjQZsePos7BwA+gQiAGYoFjKOpKCQsHDGIHS8vuhYOLgAXGRwAJDUdJRw5BRwsAyq7Llw3HwCghF2stqYFRTVcEi1bPWNmi0ibXIRlM5EBYUAAhC4kgAeMGCUtPC8MGxlREq4VBwA5EgyuZ0UXVUwNWqM3GoAYgsDcBrN2mPa+4f9z+yOS6Va4MPp3IgcMHsf6vd5aCZfCY-SF-F6OFyXUjuTzeVLpTL+HL6fphCLRKCxGRAA) to show you here @vtrofin 

So in short, using variant Constructor it will be interpreted to be an object at runtime as you can see in the image, so that is why you got the error no a function
Screenshot_2021_07_09_19_25

Comment (vtrofin) [reply src/Rules.res:2]: oh cool, definitely. thank you for the clarification @infothien

File: src/Validation.resi [hunk 2]
```diff
@@ -0,0 +1,3 @@
+type rec t
+let syncFunction: (string => bool) => t
+let asyncFunction: (string => Js.Promise.t<bool>) => t
```
Comment (gaku-sei) [inline src/Validation.resi]: I like this version better than the union as it comes with no runtime cost 😄 
Just a nitpick: would you mind calling these sync/async? (and maybe add a line return in between the different lines of code 🙏 )

Comment (infothien) [reply src/Validation.resi]: Thank you done
